File: android/core/domain/src/main/java/com/enterprise/core/domain/usecases/GetDynamicFeedUseCase.kt
package com.enterprise.core.domain.usecases

import com.enterprise.core.domain.entities.DomainResult
import com.enterprise.core.domain.entities.UIComponent
import com.enterprise.core.domain.repositories.FeedRepository
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject

/** Returns reactive stream of feed components following SSOT pattern. */
class GetDynamicFeedUseCase @Inject constructor(
    private val feedRepository: FeedRepository
) {
    operator fun invoke(): Flow<DomainResult<List<UIComponent>>> {
        return feedRepository.getDynamicFeed()
    }
}

File: ios/EnterpriseMobileApp/Sources/Core/Domain/UseCases/GetDynamicFeedUseCase.swift
import Foundation

/// Use case for fetching the dynamic CMS-driven feed with SSOT pattern.
public struct GetDynamicFeedUseCase: Sendable {
    private let feedRepository: FeedRepository

    public init(feedRepository: FeedRepository) {
        self.feedRepository = feedRepository
    }

    /// Returns a reactive stream of feed components (SSOT: local DB -> network -> local DB -> auto-emit).
    public func execute() -> AsyncStream<DomainResult<[UIComponent]>> {
        feedRepository.getDynamicFeed()
    }

    /// Forces a network refresh of the feed.
    public func refresh() async -> DomainResult<[UIComponent]> {
        await feedRepository.refreshFeed()
    }
}
